master/internal/lttb/lttb.go

```diff
@@ -0,0 +1,112 @@
+/*
+Package lttb includes a Go rewrite of Sveinn Steinarsson's reference implementation (written in
+JavaScript) of the Largest-Triangle-Three-Buckets algorithm as described in his thesis,
+"Downsampling Time Series for Visual Representation":
+
+Advisors: Jóhann Pétur Malmquist, Kristján Jónasson
+Faculty Representative: Bjarni Júlíusson
+Faculty of Industrial Engineering, Mechanical Engineering and Computer Science
+School of Engineering and Natural Sciences
+University of Iceland
+Reykjavik, June 2013
+https://skemman.is/bitstream/1946/15343/3/SS_MSthesis.pdf
+
+As far as translation makes possible, the original algorithm, variable names, and comments are
+unmodified. The accompanying tests were added after the Go rewrite. The following copyright notice
+and the MIT license pertain to the original implementation:
+
+Copyright (c) 2013 by Sveinn Steinarsson
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+package lttb
+
+import (
+	"math"
+)
+
+// Point represents a Cartesian coordinate pair.
+type Point struct {
+	X float64
+	Y float64
+}
+
+// Downsample selects the most visually significant points from a series.
+func Downsample(data []Point, threshold int) (sampled []Point) {
+	dataLength := len(data)
+	if threshold >= dataLength || threshold == 0 {
+		return data // Nothing to do
+	}
+	sampled = make([]Point, threshold)
+
+	// Bucket size. Leave room for start and end data points
+	every := float64(dataLength-2) / float64(threshold-2)
+
+	var a, nextA int // Initially a is the first point in the triangle
+	var maxArea, area float64
+
+	sampled[0] = data[a] // Always add the first point
+	sampledIndex := 1
+	for i := 0; i < threshold-2; i++ {
+		// Calculate point average for next bucket (containing c)
+		var avgX, avgY float64
+		avgRangeStart := int(float64(i+1)*every) + 1
+		avgRangeEnd := int(float64(i+2)*every) + 1
```

Comment:
question: do you end up with slightly more accurate buckets with `math.Ceil()` as oppsed to `int() + 1`. And the second would be `math.Max(avgRangeStart+1, math.Ceil())` to make sure it's not 0 width or something?

Reply:
That could be good. My biggest concern with this translation was that JavaScript's less explicit typing didn't make all of the intentions like this clear.

Reply:
makes sense just to stay as close to the reference as possible - I didn't actually read it, just an explanation online of what lttb should do (and this all seems right from that). pretty interesting stuff, something I'd never really given any thought.

Reply:
I don't have a preference in the end. I doubt it would affect the results terribly much, anyways, and this is plenty correct.